cluster-autoscaler/core/scale_up.go

```diff
@@ -374,42 +369,27 @@ func ScaleUp(context *context.AutoscalingContext, processors *ca_processors.Auto
 			Pods:      make([]*apiv1.Pod, 0),
 		}
 
-		// add list of pods which pass predicates to option
-		podsPassing, err := getPodsPassingPredicates(nodeGroup.Id())
-		if err != nil {
-			klog.V(4).Infof("Skipping node group %s; cannot compute pods passing predicates", nodeGroup.Id())
-			skippedNodeGroups[nodeGroup.Id()] = notReadyReason
-			continue
-		} else {
-			option.Pods = make([]*apiv1.Pod, len(podsPassing))
-			copy(option.Pods, podsPassing)
-		}
-
-		// update information why we cannot schedule pods for which we did not find a working extension option so far
-		podsNotPassing, err := getPodsNotPassingPredicates(nodeGroup.Id())
-		if err != nil {
-			klog.V(4).Infof("Skipping node group %s; cannot compute pods not passing predicates", nodeGroup.Id())
-			skippedNodeGroups[nodeGroup.Id()] = notReadyReason
-			continue
-		}
-
-		// mark that there is a scheduling option for pods which can be scheduled to node from currently analyzed node group
-		for _, pod := range podsPassing {
-			delete(podsRemainUnschedulable, pod)
-		}
-
-		for pod, err := range podsNotPassing {
-			_, found := podsRemainUnschedulable[pod]
-			if found {
-				podsRemainUnschedulable[pod][nodeGroup.Id()] = err
+		for _, eg := range podEquivalenceGroups {
+			samplePod := eg.pods[0]
+			if err := context.PredicateChecker.CheckPredicates(samplePod, nil, nodeInfo); err == nil {
+				// add pods to option
+				option.Pods = append(option.Pods, eg.pods...)
+				// mark pod group as (theoretically) schedulable
+				eg.schedulable = true
```

Comment:
This will not mutate the struct which is in the `podEquivalenceGroups` slice but the copy used for iteration.

Also I would prefer that `equivalenceGroup` to just be struct with groupId and slice of pods. And that we keep track of whether given group is schedulable or not and failed predicates in separate data struct. Pod equivalence regarding scheduling properties and information whether given equivalence group is schedulable on given node shape is a separate concern.

Reply:
If I am not mistaken - and there is a bug I described - can we add test which would catch that?

Reply:
Good catch! Added more assertions to existing tests (that would catch it) and also a new test for scenario where some pods trigger scale up and some don't.